### DISCUSSION TITLE
RoleName not passed to connection

### DISCUSSION THREAD

**Discussion by zhus** [fernandobatels/rsfbclient#137] [General]
My connection string is "firebird://user:pass@server:3050/control.gdb?encoding=WIN1250&&roleName=R_USER"
but RoleName not passed to connection. So I found fn build_dpb and patched with

{
 let role_name = "R_USER".to_string();
 dpb.extend(&[ibase::isc_dpb_sql_role_name as u8, role_name.len() as u8]);
 dpb.extend(role_name.bytes());
}

Is there a valid way to pass RoleName to connection?
with builder_native().with_dyn_load("fbclient.dll")

**Comment by fernandobatels** [comment]
We don´t support this yet. I will fix. 

Can you provide some example of role use? Like a expected result when some role is used.

**Reply by zhus** [reply]
A role is a database object as set of rights for user to access database so when I connect to database with no role it set to "NONE" and I can't access some tables and procedures. For example
```
let mut conn = rsfbclient::builder_native().with_dyn_load("fbclient.dll")
        .with_remote().host("server3").port(3050).db_name("control.gdb").charset(WIN_1250)
        .user("*****").pass("*****")
        .connect()?;
    let rows: Vec<(i32, i32, String)> = conn.query("SELECT * FROM PS_TASKS_SEL2_TEST(180322)", ())?;
    for row in rows { println!("{} - {} - {}", row.0, row.1, row.2); }
```
will result Error: Sql { msg: "no permission for EXECUTE access to PROCEDURE TASKS_PPN_V", code: -551 }
May be just add new function for NativeConnectionBuilder
```
    pub fn role<S: Into<String>>(&mut self, role: S) -> &mut Self {
        self.conn_conf.attachment_conf.role = role.into();
        self
    }
```
and extend build_dpb function by
```
dpb.extend(&[ibase::isc_dpb_sql_role_name as u8, role.len() as u8]);
dpb.extend(role.bytes());
```
In this case connection setup would be
```
let mut conn = rsfbclient::builder_native().with_dyn_load("fbclient.dll")
        .with_remote().host("server3").port(3050).db_name("control.gdb").charset(WIN_1250)
        .user("*****").pass("*****").role("*****")
        .connect()?;
```
Unfortunately I'm not professional programmer and absoluetly novice in Rust lang. I figured out how to make working patch under the hood but I do not know how to do it properly in whole project.

**Reply by fernandobatels** [reply]
Your response was very useful. Thanks. You can test using the `roles-names` branch  until the next version is released.

**Comment by zhus** [comment]
Thank you! All works as expected. Throwed to prod :))

**Comment by fernandobatels** [comment]
`0.22.0` version released!
